### PR TITLE
Deprecate withProcess/withProcess_ (fixes #25)

### LINKED ISSUE #25: Deprecate withProcess
I've now [twice](https://github.com/commercialhaskell/stack/issues/4888#issuecomment-505123784) been bitten by a bug around withProcess:

* Use withProcess
* Stream data from child process
* Done

Problem: this creates a really subtle race condition where the child process may close its output handles and, before actually exiting, the parent process will hit `closeProcess` and then send a SIGTERM to the child.

The real problem is that `withProcess` is confusing in how it behaves. I propose instead having two new functions:

* `withProcessTerm`, with the current behavior
* `withProcessWait`, which will never send a SIGTERM, but instead wait for the child process to exit

Adding a deprecation message referring to these two new functions would have prevented a lot of head scratching in the two cases above.

### PR DESCRIPTION
**Pull request by snoyberg**
@psibi would you mind taking a look at this?

### CODE REVIEW

File: src/System/Process/Typed.hs [hunk 1]
```diff
@@ -750,27 +754,73 @@ stopProcess = liftIO . pCleanup
...
+--
+-- @since 0.2.5.0
+withProcessWait_
+  :: ProcessConfig stdin stdout stderr
+  -> (Process stdin stdout stderr -> IO a)
+  -> IO a
+withProcessWait_ config f = bracket
+    (startProcess config)
+    stopProcess
+    (\p -> f p <* checkExitCode p)
```
Comment (psibi) [inline src/System/Process/Typed.hs]: Any reason why on this case (and in `withProcessWait`), you have chosen `<*` as compared to `finally` ?

**Comment by psibi** [comment]
But strangely the test case you have added are failing in CI: https://travis-ci.org/fpco/typed-process/jobs/550165307 (Although it does pass for me locally).

File: src/System/Process/Typed.hs (hunk 1 above)
Comment (snoyberg) [reply src/System/Process/Typed.hs]: Yes. If there was an exception in the inner block, we want that exception to propagate. Checking the exit code of the process is irrelevant when we're already about to throw a runtime exception.

**Comment by snoyberg** [comment]
I'm guessing that's a timing issue, I'll try extending the sleep and seeing what happens.

**Comment by snoyberg** [comment]
Nope, that didn't fix it at all, something else is going on.

**Comment by psibi** [comment]
@snoyberg I think it might have got to do something with GHC. I changed the resolver to `9.21` and I was able to reproduce the test failures consistently.

**Comment by snoyberg** [comment]
You're awesome @psibi!

**Comment by snoyberg** [comment]
OK, looks like the issue is that `waitForProcess` isn't interruptible, and therefore the process doesn't get stopped correctly. But only on older GHCs. I'm going to comment out these two test cases and link to this comment as an explanation for why. Thanks for helping me not be crazy here @psibi, much appreciated!

**Comment by snoyberg** [comment]
Last test failure, funnily enough, is the Stack bug that initially encouraged the creation of this PR.

**Pull request merged**